lib/govuk_app_config/govuk_logging.rb

```diff
@@ -1,32 +1,31 @@
+require "json"
 require "logstasher"
 require "action_controller"
 require_relative "rails_ext/action_dispatch/debug_exceptions"
 
 module GovukLogging
   def self.configure
-    # GOV.UK Rails applications are expected to output JSON to stdout which is
-    # then indexed in a Kibana instance. These log outputs are created by the
-    # logstasher gem.
-    #
-    # Rails applications will typically write other things to stdout such as
-    # `Rails.logger` calls or 'puts' statements. However these are not in a
-    # JSON format which causes problems for the log file parsers.
-    #
-    # To resolve this we redirect stdout to stderr, to cover any Rails
-    # writing. This frees up the normal stdout for the logstasher logs.
-    #
-    # We also disable buffering, so that logs aren't lost on crash or delayed
-    # indefinitely while troubleshooting.
-
-    # rubocop:disable Style/GlobalVars
-    $real_stdout = $stdout.clone
-    $real_stdout.sync = true
-    $stdout.reopen($stderr)
     $stdout.sync = true
-    # rubocop:enable Style/GlobalVars
 
-    # Send Rails' logs to STDERR because they're not JSON formatted.
-    Rails.logger = ActiveSupport::TaggedLogging.new(Logger.new($stderr, level: Rails.logger.level))
+    Rails.logger = Logger.new(
+      $stdout,
+      level: Rails.logger.level,
+      formatter: proc { |severity, datetime, _progname, msg|
+        begin
+          message = JSON.parse(msg)
+        rescue JSON::ParserError, TypeError => _e
+          message = msg
+        end
```

Comment:
This seems a bit gnarly 😞 so the logged message could _maybe_ be a JSON string, but _maybe_ be plain text? Is that because the log could have come from Logstasher (as a JSON string) or Rails.logger.info (as plain text)?

If so, it might be worth adding a comment here to explain that. It seems like one of those weird little details that people won't be aware of most of the time.

Reply:
I wonder if we actually need this. I'm expecting we've got it because of: https://github.com/alphagov/govuk_app_config/blob/e615e9b14966d992dd2b5a7a558a3e110d4fee5d/lib/govuk_app_config/rails_ext/action_dispatch/debug_exceptions.rb#L42-L46 but we can probably change that from being JSON formatted - I assume there isn't any apps that routinely log in JSON.

So I'd imagine we can get rid of this one bit of JSON logging and then save having to catch an exception for presumably every other log entry

Reply:
> This seems a bit gnarly 😞 so the logged message could _maybe_ be a JSON string, but _maybe_ be plain text? Is that because the log could have come from Logstasher (as a JSON string) or Rails.logger.info (as plain text)?
> 
> If so, it might be worth adding a comment here to explain that. It seems like one of those weird little details that people won't be aware of most of the time.

In short yes, the log might be json or not, and no, those logs are separated from logstashers. I totally agree on it is not pretty, maybe we can do a bit refactoring and add some comments

Reply:
> I wonder if we actually need this. I'm expecting we've got it because of:
> 
> https://github.com/alphagov/govuk_app_config/blob/e615e9b14966d992dd2b5a7a558a3e110d4fee5d/lib/govuk_app_config/rails_ext/action_dispatch/debug_exceptions.rb#L42-L46
> 
> but we can probably change that from being JSON formatted - I assume there isn't any apps that routinely log in JSON.
> So I'd imagine we can get rid of this one bit of JSON logging and then save having to catch an exception for presumably every other log entry

I believe this monkeypatch only works for errors so we still need that piece of code :(

Reply:
Ah sorry that wasn't what I meant. I'll try rephrase: Excluding [debug_exceptions.rb](https://github.com/alphagov/govuk_app_config/blob/e615e9b14966d992dd2b5a7a558a3e110d4fee5d/lib/govuk_app_config/rails_ext/action_dispatch/debug_exceptions.rb#L42-L46), what's the evidence that we have JSON strings sent to the logger?

Reply:
> Ah sorry that wasn't what I meant. I'll try rephrase: Excluding [debug_exceptions.rb](https://github.com/alphagov/govuk_app_config/blob/e615e9b14966d992dd2b5a7a558a3e110d4fee5d/lib/govuk_app_config/rails_ext/action_dispatch/debug_exceptions.rb#L42-L46), what's the evidence that we have JSON strings sent to the logger?
oh we don't, so there are 3 possibilities a) JSON b) stringify-ed JSON and c) plain text.
The parser is meant to handle stringify-ed JSON only so that they will end up nice and pretty in logit.

Reply:
> Ah sorry that wasn't what I meant. I'll try rephrase: Excluding [debug_exceptions.rb](https://github.com/alphagov/govuk_app_config/blob/e615e9b14966d992dd2b5a7a558a3e110d4fee5d/lib/govuk_app_config/rails_ext/action_dispatch/debug_exceptions.rb#L42-L46), what's the evidence that we have JSON strings sent to the logger?

Oh we don't, the msg received by logger might be JSON or plain text or stringify-ed JSON. This parser is there to make stringify-ed JSON nice and pretty in Logit, while we'll leave the other 2 as they are.

Reply:
I'm confused though why are we expecting JSON messages? who's logging them?

Reply:
Sure, I'd suggest getting rid of it. It's not like we have had JSON parsing from logging before so this is a new feature and is not keeping things backwards compatible.

Reply:
Oh yeah this one is a confusing one. Unfortunately we still need to parse the `msg`:

This monkeypatch here produces a string that contains json:
https://github.com/alphagov/govuk_app_config/blob/e615e9b14966d992dd2b5a7a558a3e110d4fee5d/lib/govuk_app_config/rails_ext/action_dispatch/debug_exceptions.rb#L42C5-L46

This govuk_logger here receives the above `msg` as a string. In most cases it is expected to be just a plain string of text, but in the case of the when it comes from the above monkeypatch, it is a string containing json.

The function wants to produce a string containing json with some extra fields. It has to parse the message first in order to avoid a situation where the end result is a string containing valid json and escaped json.

https://github.com/alphagov/govuk_app_config/blob/Log-GOVUK_REQUEST_ID-in-Json-in-Whitehall/lib/govuk_app_config/govuk_logging.rb#L12-L35

We tried removing the parsing but then these tests caught us. Yay for Tests 👍 

https://github.com/alphagov/govuk_app_config/blob/Log-GOVUK_REQUEST_ID-in-Json-in-Whitehall/spec/lib/govuk_logging_spec.rb#L85-L125

Reply:
Ah yeah, my thought was that the debug_exceptions was the only example of JSON. That was only done as JSON as a way to get everything onto one line when each line was logged.

So my proposal is: we change https://github.com/alphagov/govuk_app_config/blob/e615e9b14966d992dd2b5a7a558a3e110d4fee5d/lib/govuk_app_config/rails_ext/action_dispatch/debug_exceptions.rb#L42C5-L46 to log a block of text with new lines in it instead of JSON then we can remove the JSON parsing from this code.

Reply:
Yeah we could do that, but then the `exception_message` and `exception_class` would not be directly queriable fields in logit. They can still be queried as part of the message though...

Reply:
Sure, they've not be queryable before so it wouldn't be losing functionality. 

This is how the currently behave: 
<img width="1326" alt="Screenshot 2023-06-27 at 10 49 31" src="https://github.com/alphagov/govuk_app_config/assets/282717/21a0f1c8-9f25-440d-b5c3-15149a636a49">

Reply:
Hmm i'm seeing at least asset_manager and email-alert-api log them in queriable format. Not sure why they are different.

<img width="1355" alt="image" src="https://github.com/alphagov/govuk_app_config/assets/297901/c6dc73bf-0f43-4585-b5f1-6f2116934993">

Reply:
Yeah that's a TIL to me that they're different. I wonder if it's because they don't have tagged logging (like my example did with the UUID at the start) and thus have been recognised as JSON.

It seems to be something we must have only got with the recent replatfoming change of Kibana version too, so the parsing looks rather new

Reply:
Even if we did parse the JSON they'd be in a somewhat weird spot though and be message.exception_class rather than a top level field (as we don't merge them)

Reply:
Conceptually the `message` is the content we want to log - the main information of what is happening. It can be string or json and logit is being very clever about it whenever it happens to be json. If it's json the fields become available at the top level and are queriable.

The other fields are just meta-information about the context of the message. If we pull every field in the json-message into the meta-level, then what do we log as "the message"?

Reply:
We changed the error log output to be a full stack trace as a string, which allows us to remove the gnarly json parsing bit.

Reply:
the new error log looks like this 
<img width="1355" alt="image" src="https://github.com/alphagov/govuk_app_config/assets/297901/3f655ea9-e87c-4a70-a583-55fe88ac57f1">